server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java

```diff
@@ -0,0 +1,339 @@
+            threadPool,
+            clusterService,
+            transportService,
+            actionFilters,
+            Request::new,
+            NodeRequest::new,
+            ThreadPool.Names.FETCH_SHARD_STORE,
+            NodeStoreFilesMetadataBatch.class
+        );
+        this.settings = settings;
+        this.indicesService = indicesService;
+        this.nodeEnv = nodeEnv;
+    }
+
+    @Override
+    public void list(
+        Map<ShardId, ShardAttributes> shardAttributes,
+        DiscoveryNode[] nodes,
+        ActionListener<NodesStoreFilesMetadataBatch> listener
+    ) {
+        execute(new TransportNodesListShardStoreMetadataBatch.Request(shardAttributes, nodes), listener);
+    }
+
+    @Override
+    protected NodeRequest newNodeRequest(Request request) {
+        return new NodeRequest(request);
+    }
+
+    @Override
+    protected NodeStoreFilesMetadataBatch newNodeResponse(StreamInput in) throws IOException {
+        return new NodeStoreFilesMetadataBatch(in);
+    }
+
+    @Override
+    protected NodesStoreFilesMetadataBatch newResponse(
+        Request request,
+        List<NodeStoreFilesMetadataBatch> responses,
+        List<FailedNodeException> failures
+    ) {
+        return new NodesStoreFilesMetadataBatch(clusterService.getClusterName(), responses, failures);
+    }
+
+    @Override
+    protected NodeStoreFilesMetadataBatch nodeOperation(NodeRequest request) {
+        try {
+            return new NodeStoreFilesMetadataBatch(clusterService.localNode(), listStoreMetadata(request));
+        } catch (IOException e) {
+            throw new OpenSearchException(
+                "Failed to list store metadata for shards [" + request.getShardAttributes().keySet().stream().map(ShardId::toString) + "]",
+                e
+            );
+        }
+    }
+
+    /**
+     * This method is similar to listStoreMetadata method of {@link TransportNodesListShardStoreMetadata}
+     * In this case we fetch the shard store files for batch of shards instead of one shard.
+     */
+    private Map<ShardId, NodeStoreFilesMetadata> listStoreMetadata(NodeRequest request) throws IOException {
+        Map<ShardId, NodeStoreFilesMetadata> shardStoreMetadataMap = new HashMap<ShardId, NodeStoreFilesMetadata>();
+        for (ShardAttributes shardAttributes : request.getShardAttributes().values()) {
+            final ShardId shardId = shardAttributes.getShardId();
+            try {
+                StoreFilesMetadata storeFilesMetadata = TransportNodesListShardStoreMetadataHelper.getListShardMetadataOnLocalNode(
+                    logger,
+                    shardId,
+                    nodeEnv,
+                    indicesService,
+                    shardAttributes.getCustomDataPath(),
+                    settings,
+                    clusterService
+                );
+                shardStoreMetadataMap.put(shardId, new NodeStoreFilesMetadata(storeFilesMetadata, null));
+            } catch (IOException e) {
```

Comment:
@shiv0408  I think you would need to handle OpenSearchException also here. As you're running this loop, you would want it to complete for all the shards and now throw OpenSearchException in between.

Reply:
Updated in the latest commit.